File: leetcode/146.lru-cache.ts
/*
 * @lc app=leetcode id=146 lang=typescript
 *
 * [146] LRU Cache
 */

class ListNode_ {
    key: number;
    val: number;
    prev: ListNode_;
    next: ListNode_;
}

// @lc code=start
class LRUCache {
    head: ListNode_;
    tail: ListNode_;
    nodes: ListNode_[];
    capacity: number;
    len: number;
    keyToNode: Map<number, ListNode_>;

    constructor(capacity: number) {
        this.capacity = capacity;
        this.nodes = new Array<ListNode_>(capacity);
        this.head = new ListNode_();
        this.tail = new ListNode_();
        this.head.next = this.tail;
        this.tail.prev = this.head;
        this.len = 0;
        this.keyToNode = new Map();
    }

    private removeNode(cur: ListNode_) {
        cur.prev.next = cur.next;
        cur.next.prev = cur.prev;
        this.keyToNode.delete(cur.key);
        this.len -= 1;
    }
    private addToHead(cur: ListNode_) {
        this.head.next.prev = cur;
        cur.next = this.head.next;
        cur.prev = this.head;
        this.head.next = cur;
        this.keyToNode.set(cur.key, cur);
        this.len += 1;
    }

    get(key: number): number {
        const find = this.keyToNode.get(key);

        if (find) {
            this.removeNode(find);
            this.addToHead(find);

            return find.val;
        }

        return -1;
    }

    put(key: number, value: number): void {
        const find = this.keyToNode.get(key);
        if (find) {
            find.val = value;
            this.removeNode(find);
            this.addToHead(find);
        } else {
            const newNode = new ListNode_();
            newNode.key = key;
            newNode.val = value;
            this.addToHead(newNode);

            if (this.len > this.capacity) {
                this.removeNode(this.tail.prev)
            }
        }
    }
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
// @lc code=end

File: leetcode/146.lru-cache.cpp
/*
 * @lc app=leetcode id=146 lang=cpp
 *
 * [146] LRU Cache
 */

// @lc code=start
class DLink {
public:
    int key;
    int val;
    DLink *prev;
    DLink *next;
};
class LRUCache {
public:
    unordered_map<int, DLink*> ht;
    DLink head;
    DLink end;
    int n = 0;
    LRUCache(int capacity) {
        head.next = &end;
        head.key = -1;
        head.val = -1;
        end.prev = &head;
        end.key = -2;
        end.val = -2;
        n = capacity;
        ht.clear();
    }

    void removeFirst() {
        // cout << "remove: k:" << head.next->key << ", v:" << head.next->val << endl;
        ht.erase(head.next->key);
        head.next = head.next->next;
        head.next->prev = &head;
    }
    void removeByKey(int key) {
        ht[key]->next->prev = ht[key]->prev;
        ht[key]->prev->next = ht[key]->next;
        ht.erase(key);
    }
    void addToEnd(int key, int value) {
        DLink *curNode = new DLink(key, value, end.prev, &end);
        end.prev->next = curNode;
        end.prev = curNode;
        ht[key] = curNode;
    }
    
    int get(int key) {
        if (ht.count(key) > 0) {
            int value = ht[key]->val;
            removeByKey(key);
            addToEnd(key, value);
            return value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if (ht.count(key) > 0) {
            removeByKey(key);
            addToEnd(key, value);
        } else {
            if (ht.size() >= n) {
                removeFirst();
            }
            addToEnd(key, value);
        }

        
    }
};
/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end

